scipy/spatial/_lloyd_algorithm.py

```diff
@@ -0,0 +1,234 @@
+"""Lloyd-Max algorithm."""
+from __future__ import annotations
+
+import numpy as np
+from typing import (
+    Optional,
+    TYPE_CHECKING,
+)
+
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from scipy._lib._util import (
+        DecimalNumber, IntNumber
+    )
+
+from scipy.spatial import distance, Voronoi
+
+
+__all__ = ['lloyd_centroidal_voronoi_tessellation']
+
+
+def _l1_norm(points: np.ndarray) -> float:
+    l1 = distance.cdist(points, points, 'cityblock')
+    return np.min(l1[l1.nonzero()])
+
+
+def _lloyd_centroidal_voronoi_tessellation(
+        points: np.ndarray,
+        decay: float,
+        qhull_options: str,
+) -> np.ndarray:
+    """Lloyd-Max algorithm iteration.
+
+    Based on the implementation of Stéfan van der Walt:
+
+    https://github.com/stefanv/lloyd
+
+    which is:
+
+        Copyright (c) 2021-04-21 Stéfan van der Walt
+        https://github.com/stefanv/lloyd
+        MIT License
+
+    Parameters
+    ----------
+    points : array_like (n, d)
+        The points to iterate on.
+    decay : float
+        Relaxation decay. A positive value would move the points toward
+        their centroid, and negative value would move them away.
+        1 would move the points to their centroid.
+    qhull_options : str
+        Additional options to pass to Qhull. See Qhull manual
+        for details. (Default: "Qbb Qc Qz Qj Qx" for ndim > 4 and
+        "Qbb Qc Qz Qj" otherwise.)
+
+    Returns
+    -------
+    points : array_like (n, d)
+        The points after an iteration of Lloyd's algorithm.
+
+    """
+    new_points = np.empty_like(points)
+
+    voronoi = Voronoi(points, qhull_options=qhull_options)
+
+    for ii, idx in enumerate(voronoi.point_region):
+        # the region is a series of indices into self.voronoi.vertices
+        # remove point at infinity, designated by index -1
+        region = [i for i in voronoi.regions[idx] if i != -1]
+
+        # get the vertices for this region
+        verts = voronoi.vertices[region]
+
+        # clipping would be wrong, we need to intersect
+        # verts = np.clip(verts, 0, 1)
+
+        # move points towards centroids:
+        # Centroid in n-D is the mean for uniformly distributed nodes
+        # of a geometry.
```

Comment:
Is this true?

Reply:
If I am reading this correctly yes https://en.wikipedia.org/wiki/Centroid#Of_a_tetrahedron_and_n-dimensional_simplex

Reply:
You are reading correctly, but this is only for an n-dimensional simplex, not arbitrary polyhedra. So this might be a reasonable approximation, but it is not the real centroid. This might be a good feature to have as an option, but it is not Lloyd's original algorithm.

The `n+1` is just because of the way they've numbered the points: `0, 1, ..., n-1, n`. There are `n+1` of them.